src/packagedcode/go_mod.py

```diff
@@ -0,0 +1,199 @@
+
+# Copyright (c) 2019 nexB Inc. and others. All rights reserved.
+# http://nexb.com and https://github.com/nexB/scancode-toolkit/
+# The ScanCode software is licensed under the Apache License version 2.0.
+# Data generated with ScanCode require an acknowledgment.
+# ScanCode is a trademark of nexB Inc.
+#
+# You may not use this software except in compliance with the License.
+# You may obtain a copy of the License at: http://apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# When you publish or redistribute any data created with ScanCode or any ScanCode
+# derivative work, you must accompany this data with the following acknowledgment:
+#
+#  Generated with ScanCode and provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+#  OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+#  ScanCode should be considered or used as legal advice. Consult an Attorney
+#  for any legal advice.
+#  ScanCode is a free software code scanning tool from nexB Inc. and others.
+#  Visit https://github.com/nexB/scancode-toolkit/ for support and download.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import io
+import logging
+import re
+
+import attr
+
+
+"""
+https://golang.org/ref/mod#go.mod-files
+
+For example:
+
+module example.com/my/thing
+
+go 1.12
+
+require example.com/other/thing v1.0.2
+require example.com/new/thing v2.3.4
+exclude example.com/old/thing v1.2.3
+require (
+    example.com/new/thing v2.3.4
+    example.com/old/thing v1.2.3
+)
+require (
+    example.com/new/thing v2.3.4
+    example.com/old/thing v1.2.3
+)
+
+"""
+
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class GoMod(object):
+    # Regex expressions to parse different types of dependency
+    parse_module = re.compile(
+        r'^module\s'
+        r'(?P<module>[a-z].*)'
+    ).match
+
+    parse_module_name = re.compile(
+        r'^module\s.*/'
+        r'(?P<name>[^\s]*)'
+    ).match
+
+    parse_require = re.compile(
+        r'^require\s'
+        r'.*/(?P<name>[^\s]*)'
+        r'\s'
+        r'(?P<version>[a-z].*)'
+    ).match
+
+    parse_exclude = re.compile(
+        r'^exclude\s'
+        r'.*/(?P<name>[^\s]*)'
+        r'\s'
+        r'(?P<version>[a-z].*)'
+    ).match
+
+    parse_dep_link = re.compile(
+        r'^.*/(?P<name>[^\s]*)'
+        r'\s'
+        r'(?P<version>[a-z].*)'
+    ).match
```

Comment:
Are these going to be compiled each time a GoMod object is created?

Reply:
@majurg They will be compiled for each line in a loop. But I don't find any other way to parse the file :)

Reply:
@majurg I think these will be compile only once when the module is imported since these are class attributes. They will not be recompiled on each GoMod object creation

Reply:
@pombredanne Thanks!